components/mosquitto/examples/serverless_mqtt/main/serverless_mqtt.c

```diff
@@ -0,0 +1,368 @@
+        char *topic = message->data;
+        char *payload = message->data + topic_len_aligned;
+        if (topic_len + topic_len_aligned + 4 > size) {
+            ESP_LOGE(TAG, "Received invalid message");
+            return;
+        }
+        ESP_LOGI(TAG, "forwarding remote message: topic:%s", topic);
+        ESP_LOGI(TAG, "forwarding remote message: payload:%.*s", payload_len, payload);
+        esp_mqtt_client_publish(s_local_mqtt, topic, payload, payload_len, 0, 0);
+    }
+}
+
+static esp_err_t create_candidates(void)
+{
+    ESP_RETURN_ON_FALSE(s_state = xEventGroupCreate(), ESP_ERR_NO_MEM, TAG, "Failed to create state event group");
+    s_peer_desc_json = cJSON_CreateObject();
+    esp_err_t ret = ESP_OK;
+    juice_set_log_level(JUICE_LOG_LEVEL_INFO);
+    juice_config_t config = { .stun_server_host = CONFIG_EXAMPLE_STUN_SERVER,
+                              .stun_server_port = 19302,
+                              .cb_state_changed = juice_state,
+                              .cb_candidate = juice_candidate,
+                              .cb_gathering_done = juice_gathering_done,
+                              .cb_recv = juice_recv,
+                            };
+
+    s_agent = juice_create(&config);
+    ESP_RETURN_ON_FALSE(s_agent, ESP_FAIL, TAG, "Failed to create juice agent");
+    ESP_GOTO_ON_FALSE(juice_get_local_description(s_agent, s_buffer, MAX_BUFFER_SIZE) == JUICE_ERR_SUCCESS,
+                      ESP_FAIL, err, TAG, "Failed to get local description");
+    ESP_LOGI(TAG, "desc: %s", s_buffer);
+    cJSON_AddStringToObject(s_peer_desc_json, "desc", s_buffer);
+
+    ESP_GOTO_ON_FALSE(juice_gather_candidates(s_agent) == JUICE_ERR_SUCCESS,
+                      ESP_FAIL, err, TAG, "Failed to start gathering candidates");
+    ESP_GOTO_ON_FALSE(xEventGroupWaitBits(s_state, PEER_GATHER_DONE, pdTRUE, pdTRUE, pdMS_TO_TICKS(30000)),
+                      ESP_FAIL, err, TAG, "Failed to connect to the sync broker");
+    s_peer_desc = cJSON_Print(s_peer_desc_json);
+    ESP_LOGI(TAG, "desc: %s", s_peer_desc);
+    cJSON_Delete(s_peer_desc_json);
+    return ESP_OK;
+
+err:
+    juice_destroy(s_agent);
+    s_agent = NULL;
+    cJSON_Delete(s_peer_desc_json);
+    s_peer_desc_json = NULL;
+    return ret;
+}
+
+static void local_handler(void *args, esp_event_base_t base, int32_t id, void *data)
+{
+    switch (id) {
+    case MQTT_EVENT_CONNECTED:
+        ESP_LOGI(TAG, "local client connected");
+        break;
+    case MQTT_EVENT_DISCONNECTED:
+        ESP_LOGI(TAG, "local client disconnected");
+        break;
+
+    case MQTT_EVENT_ERROR:
+        ESP_LOGI(TAG, "local client error");
+        break;
+    default:
+        ESP_LOGI(TAG, "local client event id:%d", (int)id);
+        break;
+    }
+}
+
+static esp_err_t create_local_client(void)
+{
+    esp_err_t ret = ESP_OK;
+    esp_mqtt_client_config_t mqtt_cfg = {
+        .broker.address.uri = "mqtt://192.168.4.1:3333",
```

Comment:
can these settings be configured as Kconfig options?

Reply:
will use esp_netif get IP info to get the AP IP

Reply:
used actual IPv4 of both STA and AP